daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/ConcurrentCompiledPackages.scala

```diff
@@ -83,11 +84,15 @@ private[lf] final class ConcurrentCompiledPackages(compilerConfig: Compiler.Conf
           // map using 'computeIfAbsent' which will ensure we only compile the
           // package once. Other concurrent calls to add this package will block
           // waiting for the first one to finish.
-          if (!_packages.contains(pkgId)) {
+          if (!_signatures.contains(pkgId)) {
+            val signature = AstUtil.toSignature(pkg)
+            val signatureLookup: PartialFunction[PackageId, PackageSignature] = {
+              case `pkgId` => signature
```

Comment:
This seems fairly inefficient. You are effectively building up a linear search afaict. Any reason not to throw this stuff in a map?

Reply:
It’s constant for this one but if you are building up the whole function as a chain of `orElse` of this type of partial function, you end up doing a linear search instead of a lookup to see if a package id is in there.

Reply:
In this case lookup in  `signatureLookup orElse _signatures` is constant as `_signatures` is a Map.
I prefer to not put the signature of `pkgId` into the mutable map `_signatures` before the compilation succeed.

Reply:
Ah I see, I missed the fact that `_signatures` was a map. I thought we incrementally build up `_signatures` as a composition of a bunch of partial functions. Makes sense :+1: